### PR TITLE
15632 remotecontrol cpd import participants update

### PR DESCRIPTION
**Pull request by mbischof**
Fixed issue # : https://bugs.limesurvey.org/view.php?id=15632
New feature # : 
Changed feature # : cpd_importParticipants allows updating records
Dev: 
Dev:

### CODE REVIEW

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 1]
```diff
@@ -3115,107 +3115,105 @@ protected function _checkSessionKey($sSessionKey)
...
      * [[0] => ["email"=>"dummy-02222@limesurvey.com","firstname"=>"max","lastname"=>"mustermann"]]
+     * @param bool $update
      * @return array with status
      */
-    public function cpd_importParticipants($sSessionKey, $aParticipants)
+    public function cpd_importParticipants($sSessionKey, $participants, $update = false)
     {
-
         if (!$this->_checkSessionKey($sSessionKey)) {
-            return array('status' => 'Invalid session key');
```
Comment (Shnoulle) [inline application/helpers/remotecontrol/remotecontrol_handle.php]: No, all error return an arry with status. Don't update this part

File: application/helpers/remotecontrol/remotecontrol_handle.php [hunk 2]
```diff
@@ -3115,107 +3115,105 @@ protected function _checkSessionKey($sSessionKey)
...
+                    $model->participant_id = Participant::gen_uuid();
                 }
+            }
 
-                // Write to database if record not exists
-                if (empty($arRecordExists)) {
-                    // save participant to database
-                    Participant::model()->insertParticipantCSV($aData);
+            $scenario = $model->getScenario(); // insert or update
+            if ($scenario == 'update' && $update === false) {
```
Comment (Shnoulle) [inline application/helpers/remotecontrol/remotecontrol_handle.php]: Strange to use getScenario :)

Why not , maybe more simple to understand 

````
if (!$model) {
    // Create
} elseif (!$update) {
    continue;
}
````

**Comment by olleharstedt** [comment]
Looks like lots of tests fail for some reason. I will run them locally and check.

**Comment by olleharstedt** [comment]
The test fails because there are too many failed login attempts.

**Comment by olleharstedt** [comment]
Nope, wrong. The test fails because the default user is overwritten from "admin" to "remotecontroladmin".

**Comment by mbischof** [comment]
this password? https://github.com/LimeSurvey/LimeSurvey/blob/master/tests/helpers/RemoteControlTest.php#L33

**Comment by Shnoulle** [comment]
Scrutinizer is more happy now :). 

Else : can you review the website and the email ?

File: tests/helpers/remotecontrol/BaseTest.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    /* @var remotecontrol_handle*/
+    public $handler;
+
+    const REMOTECONTROL_USERNAME = 'admin';
+    const REMOTECONTROL_PASSWORD = 'password';
+    const REMOTECONTROL_EMAIL    = 'admin@test.de';
+
+    public static function setUpBeforeClass()
+    {
+        Yii::import('application.helpers.common_helper', true);
```
Comment (olleharstedt) [inline tests/helpers/remotecontrol/BaseTest.php]: These lines are duplicated - they are already here: https://github.com/LimeSurvey/LimeSurvey/blob/master/tests/TestHelper.php#L23

**Comment by olleharstedt** [comment]
I'm guessing you want this merged into new branch 3.x-LTS, not master (which is now LS4)?

**Comment by mbischof** [comment]
thats correct

**Comment by maziminke** [comment]
We also need this fix to be available at Limesurvey 4 later.

**Comment by olleharstedt** [comment]
> We also need this fix to be available at Limesurvey 4 later.

No problem.

**Comment by olleharstedt** [comment]
> thats correct

:+1: 

Please note that I won't merge until the issues are resolved (code duplication, in this case).

**Comment by Shnoulle** [comment]
https://tools.ietf.org/html/rfc2606 or limesurvey.org for example . limesurvey.example can be good too :)

**Comment by olleharstedt** [comment]
Can you adjust the comments by Denis, please?

File: application/helpers/remotecontrol/remotecontrol_handle.php (hunk 2 above)
Comment (Shnoulle) [reply application/helpers/remotecontrol/remotecontrol_handle.php]: Still think i prefer the elseif method. What it's your opinion @olleharstedt ?

Comment (olleharstedt) [reply application/helpers/remotecontrol/remotecontrol_handle.php]: No strong opinion. So will merge.

**Comment by olleharstedt** [comment]
FYI, this PR has been cherry-picked into the master branch.

**Pull request merged**